### PR TITLE
Add endpoints for votes and also pytest for votes

### PR DESCRIPTION
**Pull request by thoo**

### CODE REVIEW

**Comment by johnmarcampbell** [comment]
I haven't delved into looking at the code yet, but I noticed you've made three commits with the same commit message. I'd suggest using a different commit message for each commit, so it's easier to tell what each commit does. If you want to change a previous message, or add new changes to the last commit, you can use `git commit --amend`. However, I'm not sure if that plays nicely with commits you've already pushed.

**Comment by thoo** [comment]
@johnmarcampbell  I just squashed a few commits which are redundant.

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Congress [![Build Status](https://travis-ci.org/thedataincubator/congress-api.svg?branch=master)](https://travis-ci.org/thedataincubator/congress-api)
- This is the API address for propublica: [propublica link] (https://projects.propublica.org/api-docs/congress-api/)
+ This is the API address for propublica(https://projects.propublica.org/api-docs/congress-api/):
```
Comment (johnmarcampbell) [inline README.md]: I think this could be expressed more succinctly by combining it with the sentence below as:
"This repo is a python package to access the [Propublica API](https://projects.propublica.org/api-docs/congress-api/)."

And a minor point for the future: best practice is to use separate commits for unrelated things. So a quick edit to the README.md file should be committed separately from adding a bunch of code about the votes endpoint.

File: congress/votes.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
+from .base import Client, SingleListParser, BaseValidator, BaseParser
+
+class Votes:
+    """Class for interacting with `Votes` endpoint of the Propublica Congress API
```
Comment (johnmarcampbell) [inline congress/votes.py:5]: Add a blank line below this. The [python docstring guidelines](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) mention that: "The summary line may be used by automatic indexing tools; it is important that it fits on one line and is separated from the rest of the docstring by a blank line."

If you want to add more detail than a one line summary, you can include it after the blank line.

File: congress/votes.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    ----------
+    api_key : str
+        The Propublica API key to use for authentication
+    """
+
+    def __init__(self, api_key):
+        self._api_key = api_key
+
+    def get_recent_votes(self,chamber,offset=0):
+        """This request returns the 20 most recent results,
```
Comment (johnmarcampbell) [inline congress/votes.py]: - Same as above, put the summary on a single line. I'll stop mentioning at after this, but this should be the same for all the docstrings (including the blank line after).

- Docstrings for functions/methods should read as a command, so you can shorten this to "Return the 20 most recent results, sorted by date and roll call number.", with similar edits for other functions

File: congress/votes.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        -------
+        dict
+            A dictionary containing the json response
+
+        """
+        
+        client = Client(self._api_key, 
+                        BaseParser(),
+                        BaseValidator())
+        return client.get(congress,chamber,\
```
Comment (johnmarcampbell) [inline congress/votes.py]: - You can remove the trailing backslashes here and elsewhere. The trailing comma tells python that the statement continues on the next line.
- Here and elsewhere, consider adding a space after the commas that separate arguments. White space can make the code more readable.

File: congress/votes.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            A dictionary containing the json response
+        """
+
+        client = Client(self._api_key, 
+                        BaseParser(),
+                        BaseValidator())
+        return client.get(chamber, "votes/recent.json",offset=offset)
+
+    def get_rollcall_votes(self,congress,chamber,ses_no,rcall_no):
+        """
```
Comment (johnmarcampbell) [inline congress/votes.py:38]: Add a summary line here. Same for a few other functions that don't have them.

**Review by johnmarcampbell — CHANGES_REQUESTED** [review]
Great work! All the code looks good and the tests all pass. The documentation looks good, and my only comments are a few stylistic tweaks. (We'll eventually be using Sphinx to auto-generate documentation, so things that seem sort of pedantic here are just to make it easier for Sphinx to recognize things.) I made some comments below, but I'll summarize here:

- All docstrings should have a short summary that fits on one line and is followed by a blank line.
- Summaries for functions/methods should read like a command, i.e., "Return this" instead of "This function returns this"

Thanks!

**Comment by thoo** [comment]
Thanks @johnmarcampbell . I think I fixed all. Let me know if I miss something or misunderstood something.

**Review by johnmarcampbell — APPROVED** [review]

**Comment by johnmarcampbell** [comment]
Looks good! One minor thing I didn't mention, it's best if the summary line fits within 80 characters or so. The docstrings for `recent_expl` and `recent_votes_expl` are pretty long for a "summary." I'd suggest something like:
```
"""
Return personal explanations for missed or mistaken votes

These explanations can refer to a single vote or to multiple votes.

Parameters
----------
...
```

**Comment by johnmarcampbell** [comment]
Approving the changes. Thanks!
@ZachGlassman Do you want to look at the PR?

**Comment by ZachGlassman** [comment]
@johnmarcampbell feel free to assign me to review.

File: congress/__init__.py [hunk 6]
```diff
@@ -1,2 +1,3 @@
 from .base import Client, BaseParser, BaseValidator
 from .congress import Congress
+from .votes import Votes
```
Comment (ZachGlassman) [inline congress/__init__.py]: Please do use the new line at end of file convention.  This can be enabled in most text editors

**Comment by johnmarcampbell** [comment]
@ZachGlassman I'm still a little uncertain about the right PR workflow. After @thoo made the changes I requested, I naturally clicked "Approve Changes". But this closed the PR without merging. And then you had to reopen the PR. This seems clunky. Is there a better way?

**Comment by ZachGlassman** [comment]
When approving changes, it should not automatically close the PR, although its possible I put some odd settings in there.

File: congress/__init__.py (hunk 6 above)
Comment (thoo) [reply congress/__init__.py]: Done. Thanks.

File: congress/votes.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+            A dictionary containing the json response
+
+        """
+        
+        client = Client(self._api_key, 
+                        BaseParser(),
+                        BaseValidator())
+        return client.get(congress,chamber, 
+                          "sessions",ses_no,"votes", 
+                          rcall_no+".json")
```
Comment (ZachGlassman) [inline congress/votes.py]: Lets use `"{}.json".format(rcall_no)` and please fix spacing

Comment (ZachGlassman) [reply congress/votes.py]: this applies in a few places

File: congress/votes.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+            A dictionary containing the json response
+        """ 
+        
+        client = Client(self._api_key, 
+                        BaseParser(),
+                        BaseValidator())
+        return client.get(congress,"explanations", 
+                          "votes.json",offset=offset)
+        
+    def recent_expl_bycat(self,congress,category,offset=0):
```
Comment (ZachGlassman) [inline congress/votes.py]: spacing

File: congress/votes.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        """
+        Return personal explanations for missed or mistaken votes in the Congressional Record 
+        
+        Parameters
+        ----------
+        congress : str
+            From 110 to 115 
+        category : str
+            "voted-incorrectly","official-business","personal"
```
Comment (ZachGlassman) [inline congress/votes.py:192]: if there are only a few categories, this is where you can use a validator

**Review by ZachGlassman — CHANGES_REQUESTED** [review]
a few formatting things, the big thing is to add in some validation and then tests for that

File: congress/votes.py (hunk 9 above)
Comment (thoo) [reply congress/votes.py:192]: @ZachGlassman  Should I also write validators for `congress` and `chamber` ? Or are they too obvious? Thanks.

Comment (ZachGlassman) [reply congress/votes.py:192]: Wherever it makes sense, please add them

**Comment by ZachGlassman** [comment]
@thoo is this ready again for review?

**Comment by thoo** [comment]
@ZachGlassman  Thank Zach. Could you please take a look?

File: test/test_votes.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    congress_no = '105'
+    chamber = 'senate'
+    vote_type = 'party'
+
+    json_sample_name = 'test/api_response_samples/get_votesbytype.json'
+    mock_json = read_json(json_sample_name)
+
+    api_url = client.build_url(congress_no, chamber, "votes",
+                               "{}.json".format(vote_type))
+    print(api_url)
```
Comment (ZachGlassman) [inline test/test_votes.py:75]: If the print statement is not necessary, lets remove it

**Comment by johnmarcampbell** [comment]
`Fix syntax with Flask8` should maybe be `Fix syntax with Flake8`?

File: congress/votes.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        chamber : str
+            `house`, `senate`
+        vote_type : str
+            `missed`, `party`, `loneno` or `perfect`
+
+        Returns
+        -------
+        pandas.DataFrame
+        """
+        if vote_type not in ['missed', 'party', 'loneno', 'perfect']:
```
Comment (ZachGlassman) [inline congress/votes.py:84]: we should be doing this in the validator, but correct idea.  I recommend writing a validator which checks membership in a set which can probably be used for all of these sorts of errors.

**Review by ZachGlassman — CHANGES_REQUESTED** [review]
a few more changes, almost there!

**Comment by thoo** [comment]
@ZachGlassman  Thanks for the comments. For the validator, is there any place I can learn more about it? [Pandas](https://pandas.pydata.org/pandas-docs/stable/contributing.html#test-driven-development-code-writing) has a great document about testing but couldn't find much for validation. I think I might be looking at the wrong section.

**Comment by ZachGlassman** [comment]
The client is instantiated with a Validator object, you probably want to create one which validates in the way you want, namely to check that a certain argument is within a set.
